### PR TITLE
Expanding fr

### PR DESCRIPTION
**Pull request by willmcgugan**
Expand "auto" container if any child has a relative  dimension (`fr` or `%`).

Fix margins not being respected when containers have auto dimensions.

Fixes https://github.com/Textualize/textual/issues/2220

### CODE REVIEW

File: src/textual/css/styles.py [hunk 1]
```diff
@@ -384,6 +384,21 @@ def auto_dimensions(self) -> bool:
...
+        return width is not None and width.unit in (
+            Unit.FRACTION,
+            Unit.PERCENT,
+        )
+
+    @property
+    def is_relative_height(self) -> bool:
+        """Does the node have a relative width?"""
+        height = self.height
+        return height is not None and height.unit in (Unit.FRACTION, Unit.PERCENT)
```
Comment (rodrigogiraoserrao) [inline src/textual/css/styles.py:400]: Aren't `w`, `h`, `vh`, and `vw`, relative units too?

File: src/textual/layouts/horizontal.py [hunk 2]
```diff
@@ -23,10 +23,23 @@ def arrange(
         x = max_height = Fraction(0)
         parent_size = parent.outer_size
 
+        child_styles = [child.styles for child in children]
+        box_margins = [styles.margin for styles in child_styles]
```
Comment (rodrigogiraoserrao) [inline src/textual/layouts/horizontal.py:27]: Is there any reason you didn't do
`box_margins = [child.styles.margin for child in children]`?

File: src/textual/layouts/horizontal.py [hunk 3]
```diff
@@ -23,10 +23,23 @@ def arrange(
...
+                max(margin1.right, margin2.left)
+                for margin1, margin2 in zip(box_margins, box_margins[1:])
+            )
+            if len(box_margins) > 1:
+                total_margin += box_margins[0].left + box_margins[-1].right
+        else:
+            total_margin = 0
+
         box_models = resolve_box_models(
             [child.styles.width for child in children],
```
Comment (rodrigogiraoserrao) [inline src/textual/layouts/horizontal.py:39]: Did you mean to reuse `child_styles` here?

File: tests/snapshot_tests/snapshot_apps/auto_fr.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        margin: 1 2;    
+        background: $primary;
+    }
+    """
+
+    def compose(self) -> ComposeResult:
+        with Widget(id="container"):
+            yield Label("Hello one line", id="top")
+            yield Widget(id="child")
+            yield Label("Two\nLines with 2x2 margin", id="bottom")
```
Comment (rodrigogiraoserrao) [inline tests/snapshot_tests/snapshot_apps/auto_fr.py:35]: May not be super relevant but the margin set in the CSS is actually 1x2!

**Review by rodrigogiraoserrao — COMMENTED** [review]
Looking forward to using the new mechanism to handle fr units!

**Review by davep — APPROVED** [review]

File: src/textual/css/styles.py (hunk 1 above)
Comment (willmcgugan) [reply src/textual/css/styles.py:400]: `w` and `h` become percent units.

`vh` and `vw` aren't relative to the container, unless the container happens to be screen. Maybe the name could be clarified?

File: tests/snapshot_tests/snapshot_apps/auto_fr.py (hunk 4 above)
Comment (willmcgugan) [reply tests/snapshot_tests/snapshot_apps/auto_fr.py:35]: We spotted

File: src/textual/layouts/horizontal.py (hunk 2 above)
Comment (willmcgugan) [reply src/textual/layouts/horizontal.py:27]: Yes, an optimization I forgot to port from horizontal.

File: src/textual/css/styles.py (hunk 1 above)
Comment (rodrigogiraoserrao) [reply src/textual/css/styles.py:400]: A clarification is always welcome, in my opinion.

**Pull request merged**